x-pack/plugins/files/server/file_share_service/generate_share_token.ts

```diff
@@ -0,0 +1,57 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import crypto from 'crypto';
+
+/**
+ * This code is adapted and borrowed from
+ * https://github.com/sindresorhus/crypto-random-string/blob/25f8930282bc3803d03b9080e710db8426430b63/core.js
+ */
+
+const CHAR_SET = 'abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789';
+const CHAR_SET_LENGTH = CHAR_SET.length;
+
+// We are going to read as 16 bits numbers at a time which means this is the largest
+// number we are going to see.
+const MAX_16_BIT_INT = Math.pow(2, 16) - 1; // OR 0x10000 - 1
+
+// The biggest number we can use for modulo given the charset to ensure good distribution
+// This number works because we know CHAR_SET_LENGTH fits into it CHAR_SET_LENGTH number of times
+const MAX_USABLE_VALUE = Math.floor(MAX_16_BIT_INT / CHAR_SET_LENGTH) * CHAR_SET_LENGTH;
+
+// Tokens must be 40 chars long
+const TOKEN_LENGTH = 40;
+// How much entropy we will generate for each pass
+const ENTROPY_LENGTH = 2 * Math.floor(1.1 * TOKEN_LENGTH);
+
+/**
+ * Generate 40 random characters of a pre-checked charset for share tokens.
+ *
+ * Samples:
+ *
+ * XBWUuHRZsxXZgS03VUp2hSifmCoF4zznsNOehTea
+ * 40RIzynCs2OFTkYC2fy9vfLgYhxjmwzYrJvGj2Ih
+ * ZFM1vFSz6k8k9QU4Ve5mu0PbcBi3rwV4BykbcrDU
+ * 2oweE4dmWT7warVYA59l0thCDWnIMvUsAC9yCWyQ
+ * q8R4LneRCBhcsCNKgcebZRh1Mh4m746GNTuifmAA
+ * kpUzKyynu8uXDtek2Xg5yJRwPkw37Cd1VaUezNZ2
+ */
+export function generateShareToken(): string {
+  let token = '';
+  while (token.length < TOKEN_LENGTH) {
+    // It is highly unlikely we will call this function more than once per token
+    const entropy = crypto.randomBytes(ENTROPY_LENGTH);
+    let entropyPointer = 0;
+    while (entropyPointer < entropy.length && token.length < TOKEN_LENGTH) {
+      const value = entropy.readUInt16LE(entropyPointer);
+      entropyPointer += 2; // Move ahead 2 bytes to the next number
+      if (value > MAX_USABLE_VALUE) continue;
+      token += CHAR_SET[value % CHAR_SET_LENGTH];
+    }
+  }
```

Comment:
Maybe to simplify we can just Base64 encode the buffer? UUIDs provide up to 16 bytes of entropy, we just generate at least that or more, say 24 bytes, and Base64 encode:

```suggestion
export function generateShareToken(): string {
  return crypto.randomBytes(24)
    .toString('base64')
    // Replace non-alphanumeric characters
    .replace(/^[a-z0-9]/i, () => CHAR_SET[Math.floor(Math.random() * CHAR_SET.length)]);
}
```

Reply:
Given your comments perhaps it is better to not have this kind of code in Kibana (it is more "lib" level code). I just enjoyed implementing the algo myself and generating a the kind of random token I think we would want for share tokens.

I think, for this use case, we could just use the `randomUUID` function from crypto https://nodejs.org/api/crypto.html#cryptorandomuuidoptions

Then strip out `-` chars. Something like:

`crypto.randomUUID().randomUUID().replace(/-/g, '')`

I think this is likely good enough even though it is hex charset, but then we are not working directly with `randomBytes` which may make the code simpler.

WDYT?

Reply:
I wasn't suggesting to use `Math.random()`, my suggestion uses `crypto.randomBytes`:

<img width="685" alt="image" src="https://user-images.githubusercontent.com/82822460/177333678-89e086df-fdfa-4af2-b559-d1922391d5fd.png">

The `Math.random()` at the end is used to replace characters, that are not `[a-zA-Z0-9]`, i.e. `/`, `+`, and `=` (so the ID is URL-friendly):

<img width="660" alt="image" src="https://user-images.githubusercontent.com/82822460/177333859-7e98f572-7d47-43e1-9166-10623739984d.png">

(which is less than 2% of the charset, so we are not losing much "randomness  rigor").

---

> The current algo, pretty much copy of https://github.com/sindresorhus/crypto-random-string/blob/25f8930282bc3803d03b9080e710db8426430b63/core.js#L9, provides a (theoretically) perfectly equal distribution of the charset

I'm not sure I understand why we need it, as `crypto.randomBytes` already gives us cryptographically well-built random data. All we need is to convert it to a string. But the way `crypto-random-string` library converts it to a string is weird, IMO, or at least overcomplicated.

---

> WDYT?

1. I like the idea of using `crypto.randomUUID`, in principle. Only problem is the IDs will be Base16 encoded, which IMO would be nicer to have a higher encoding base.

2. I like Base64, in the function I gave as example. The only problem there are the `/`, `+`, and `=` characters, which are replaced by the trailing `.replace()` call.

3. Alternatively, we can encode in Base36, with any extra replacements:

```ts
const generateKey = () =>
  [...crypto.randomBytes(24)].map(c => c.toString(36)).join('')
```

We can put it in a separate package, but if the algorithm is as simple as any one of the three above, we can keep it where it is now.

Reply:
I think it for generating random strings that have a role in security it's best to avoid math.random entirely, since it does not play a necessary role IMO.

Reply:
`Math.random()` is not used to generate randomness here. It is used to replace 63rd and 64th standard Base64 encoding characters by something URL-friendly; that could also be done in many other ways without `Math.random()`. Like, we can replace them by `-` and `_` to get base64url format:

```ts
crypto.randomBytes(24)
    .toString('base64')
    .replace('+', '-')
    .replace('/', '_');
```

Reply:
BTW, maybe that is the way to go (instead of Base36):

```ts
const generateKey = () =>
  crypto.randomBytes(24)
    .toString('base64')
    .replace('+', '-')
    .replace('/', '_');

```